### PR TITLE
Messaging: Implement Private Messaging

### LINKED ISSUE #203: Internal Messaging?
With more users coming on board, I'm realizing that it'd be really useful ping/message another user. I'm wondering if it'd be useful at all to allow people to send each other messages within the system? We can't see other users contact information and have no way of contacting them at the moment.

### PR DESCRIPTION
**Pull request by nlewis2142**
Closes #203

Users currently have no way to contact each other because the system does not share personal information. This will allow users to message eachother.

There's quite a few files, so I would appreciate a code review to catch things I missed. As per Kohana, I tried to use HTML::entities() to escape XSS as well as encode any characters that wouldn't display correctly, but help check that XSS is indeed properly escaped.

Routing: `<account>/messages(/<action>(/<id>))`

<a href="https://picasaweb.google.com/lh/photo/91uIZ7d-xN2RgIgIaWFOAsjBVG_0f5B7oOHmZugTIg0?feat=embedwebsite"></a>

<a href="https://picasaweb.google.com/lh/photo/xIprcPf32Pw3eP26XKVW_MjBVG_0f5B7oOHmZugTIg0?feat=embedwebsite"></a>

<a href="https://picasaweb.google.com/lh/photo/hZJ7a7BAYjL26GadE4pM-sjBVG_0f5B7oOHmZugTIg0?feat=embedwebsite"></a>

<a href="https://picasaweb.google.com/lh/photo/7KKIWvkw1KFBwHCqaKCF6sjBVG_0f5B7oOHmZugTIg0?feat=embedwebsite"></a>

<a href="https://picasaweb.google.com/lh/photo/iB0xEo2cRpdXBN6MmWUDmMjBVG_0f5B7oOHmZugTIg0?feat=embedwebsite"></a>

### CODE REVIEW

File: application/classes/controller/user.php [hunk 1]
```diff
@@ -109,6 +113,31 @@ public function action_index()
 				->bind('owner', $this->owner)
 				->bind('account', $this->visited_account);
 			$gravatar_view = TRUE;
+
+			$messages = ORM::factory('message')
+				->where('recipient_id', '=', $this->user->id)
+				->order_by('read', 'asc')
+				->order_by('timestamp', 'desc')->limit(5)
+				->find_all()
+				->as_array();
```
Comment (ekala) [inline application/classes/controller/user.php]: Perhaps move this code block (message fetching) to utility method say `get_messages()` in `Model_User`. That way, the code becomes testable by way of unit tests.

File: themes/default/views/pages/message/inbox.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
+<?php
+function soft_truncate($string, $length = 200)
+{
+	$string = str_replace("\n", ' ', $string);
+	if (strlen($string) > $length)
+	{
+		$string = substr(wordwrap($string, $length), 0, strpos($string, "\n"));
+	}
+	return $string;
+}
```
Comment (ekala) [inline themes/default/views/pages/message/inbox.php]: Extend Kohana's Text (`system/classes/text.php`) helper class to include this function. As an example, see `application/classes/html.php`

Once again, makes it easier to test the same via unit tests.

File: themes/default/media/css/styles.css [hunk 3]
```diff
@@ -2274,6 +2275,84 @@ article.modal div.search ul.dual-buttons {
...
+}
+.message .message-content .button-reply {
+  float: right;
+}
+
+.messages_link {
+  float: right;
+  line-height: 2.5em;
+  padding: 0 10px;
+}
```
Comment (ekala) [inline themes/default/media/css/styles.css:2355]: The styling changes also need to be reflected in `styles.less` - `themes/default/media/css/less/styles.less` Looping in @brandonrosage

File: themes/default/views/pages/message/inbox.php (hunk 2 above)
Comment (nlewis2142) [inline themes/default/views/pages/message/inbox.php]: Digging into the Text class made me realize there was a similar function I could use. Kohana has all kinds of not-so-well documented features!

File: application/classes/controller/message.php [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+		// CHECK: Are we reading a specific message?
+		if ($this->id)
+			return $this->action_inbox_read();
+
+		$this->messages = ORM::factory('message')
+			->where('recipient_id', '=', $this->user->id)
+			->order_by('read', 'asc')
+			->order_by('timestamp', 'desc')
+			->find_all()
+			->as_array();
```
Comment (ekala) [inline application/classes/controller/message.php:41]: Perhaps move this code block to a model

Comment (nlewis2142) [inline application/classes/controller/message.php:41]: @ekala Does that match up with the coding style? From what I could tell it was standard to always call the factory function in the controller. It seems a bit strange to have the model initialize itself (which is what `ORM::factory()` does, right?), even if it's through a static function.

**Comment by nlewis2142** [comment]
@ekala @69mb I'd like to get this merged into `develop`. Any other comments you have or changes I should make?

File: themes/default/media/css/styles.css (hunk 3 above)
Comment (nlewis2142) [inline themes/default/media/css/styles.css:2355]: Can we copy/paste the differences in CSS or does `styles.less` use different styling?

File: install/sql/updates/2012_07_06_00_messages.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS `messages` (
+  `id` BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
+  `recipient_id` BIGINT UNSIGNED NOT NULL,
+  `sender_id` BIGINT UNSIGNED NOT NULL DEFAULT 0,
+  `subject` VARCHAR(255) NOT NULL DEFAULT '',
+  `message` TEXT NOT NULL DEFAULT '',
+  `read` TINYINT(1) UNSIGNED NOT NULL DEFAULT 0
```
Comment (69mb) [inline install/sql/updates/2012_07_06_00_messages.sql]: Missing ',' at the end of this line

File: themes/default/media/css/styles.css (hunk 3 above)
Comment (69mb) [inline themes/default/media/css/styles.css:2355]: Copy paste is ok

**Pull request closed without merging**